sonar-java-plugin/src/main/java/org/sonar/plugins/java/JavaSonarWayProfile.java

```diff
@@ -93,40 +98,48 @@ private static String readResource(URL resource) {
     }
   }
 
-  @SuppressWarnings("unchecked")
   @VisibleForTesting
   static Set<RuleKey> getSecurityRuleKeys(boolean sonarSecurityBefore78) {
+    String ruleKeysMethod = sonarSecurityBefore78 ? "getRuleKeys" : "getSecurityRuleKeys";
+    return getExternalRuleKeys(SECURITY_RULES_CLASS_NAME, ruleKeysMethod, "security", sonarSecurityBefore78);
+  }
+
+  @VisibleForTesting
+  static Set<RuleKey> getDataflowBugDetectionRuleKeys() {
+    return getExternalRuleKeys(DBD_RULES_CLASS_NAME, DBD_RULE_KEYS_METHOD_NAME, "dataflow bug detection", false);
+  }
+
+  @SuppressWarnings("unchecked")
+  @VisibleForTesting
+  static Set<RuleKey> getExternalRuleKeys(String className, String ruleKeysMethod, String rulesCategory, boolean sonarSecurityBefore78) {
     try {
-      Class<?> javaRulesClass = Class.forName("com.sonar.plugins.security.api.JavaRules");
-      String ruleKeysMethod = sonarSecurityBefore78 ? "getRuleKeys" : "getSecurityRuleKeys";
+      Class<?> javaRulesClass = Class.forName(className);
       Method getRuleKeysMethod = javaRulesClass.getMethod(ruleKeysMethod);
       Set<String> ruleKeys = (Set<String>) getRuleKeysMethod.invoke(null);
+      Method getRepositoryKeyMethod = javaRulesClass.getMethod(GET_REPOSITORY_KEY);
       String repositoryKey;
       if (sonarSecurityBefore78) {
         repositoryKey = CheckList.REPOSITORY_KEY;
       } else {
-        Method getRepositoryKeyMethod = javaRulesClass.getMethod("getRepositoryKey");
         repositoryKey = (String) getRepositoryKeyMethod.invoke(null);
       }
       return ruleKeys.stream().map(k -> RuleKey.of(repositoryKey, k)).collect(Collectors.toSet());
-
     } catch (ClassNotFoundException e) {
-      LOG.debug("com.sonar.plugins.security.api.JavaRules is not found, no security rules added to Sonar way java profile: " + e.getMessage());
+      LOG.debug(String.format("%s is not found, no %s rules added to Sonar way java profile: %s", className, rulesCategory, e.getMessage()));
     } catch (NoSuchMethodException e) {
-      LOG.debug("Method is not found, no security rules added to Sonar way java profile: " + e.getMessage());
+      LOG.debug(String.format("Method is not found, no %s rules added to Sonar way java profile: %s", rulesCategory, e.getMessage()));
     } catch (IllegalAccessException e) {
-      LOG.debug("[IllegalAccessException] no security rules added to Sonar way java profile: " + e.getMessage());
+      LOG.debug(String.format("[IllegalAccessException] no %s rules added to Sonar way java profile: %s", rulesCategory, e.getMessage()));
     } catch (InvocationTargetException e) {
-      LOG.debug("[InvocationTargetException] no security rules added to Sonar way java profile: " + e.getMessage());
+      LOG.debug(String.format("[InvocationTargetException] no %s rules added to Sonar way java profile: %s", rulesCategory, e.getMessage()));
```

Comment:
I think it would be a lot nicer to combine all of these into a single multi-catch block (`catch (ClassNotFoundException|NoSuchMethodExcpetion|IllegalAccessException|InvocationTargetException e)` and unify the message to `"[%s] No %s rules added to Sonar way java profile: %s"`. This slightly changes the message for the first two cases, but that should not matter - we don't lose any information because the class name is still present in the exception message. And it just feels more consistent to have the same message in all cases. And most importantly we won't have four catch blocks that all log basically the same message.

As an added bonus, this should get the coverage to 100%.

Reply:
Makes a lot of sense, thanks for the suggestion! At first I didn't want to change the existing behavior for security rules but indeed I think this makes the code much clearer.